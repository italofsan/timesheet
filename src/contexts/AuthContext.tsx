import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useEffect,
  useState,
} from "react";
import { auth, firebase } from "../services/firebase";
import { errorMessage } from "../components/Messages";

type User = {
  id: string;
  name: string;
};

type AuthContextType = {
  user: User;
  signInWithGoogle: () => Promise<void>;
  signOut: () => Promise<void>;
  isSigned: boolean;
  setIsSigned: Dispatch<SetStateAction<boolean>>;
};

type AuthContextProviderProps = {
  children: ReactNode;
};

export const AuthContext = createContext({} as AuthContextType);

export function AuthContextProvider(props: AuthContextProviderProps) {
  const [user, setUser] = useState<User>({
    id: "",
    name: "",
  });
  const [isSigned, setIsSigned] = useState<boolean>(false);

  function loadStorageData() {
    const storageIsSigned = localStorage.getItem("@SIGNED");

    if (storageIsSigned) {
      setIsSigned(JSON.parse(storageIsSigned));
    }
  }

  useEffect(() => {
    loadStorageData();
    const unsubscribe = auth.onAuthStateChanged((user) => {
      if (user) {
        const { displayName, uid } = user;

        if (!displayName) {
          throw new Error("Missing information from Google Account.");
        }

        setUser({
          id: uid,
          name: displayName,
        });
        setIsSigned(true);
        localStorage.setItem("@SIGNED", JSON.stringify(true));
      }
    });

    return () => {
      unsubscribe();
    };
  }, []);

  async function signInWithGoogle() {
    const provider = new firebase.auth.GoogleAuthProvider();
    try {
      const result = await auth.signInWithPopup(provider);
      if (result.user) {
        const { displayName, uid } = result.user;
        if (!displayName) {
          throw new Error("Missing information from Google Account.");
        }
        setUser({
          id: uid,
          name: displayName,
        });
      }
    } catch (error) {
      console.log(error);
      errorMessage("Não foi possível realizar Login!");
    }
  }

  async function signOut() {
    await auth.signOut();
    setUser({} as User);
    localStorage.clear();
    setIsSigned(false);
  }

  return (
    <AuthContext.Provider
      value={{
        user,
        signInWithGoogle,
        signOut,
        isSigned,
        setIsSigned,
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );
}
