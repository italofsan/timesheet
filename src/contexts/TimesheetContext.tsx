import {
  createContext,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from "react";
import { errorMessage, successMessage } from "../components/Messages";
import { DataTimesheetInput, DataTimesheet } from "../utils/types";
import {
  calcUndertimeHours,
  calcOvertimeHours,
  calcWorkedHours,
  sumHours,
} from "../utils/functions/mathHoursFunctions";
import { isTimesheetExists } from "../utils/functions/timesheetFunctions";
import { database } from "../services/firebase";
import { useAuth } from "../hooks/useAuth";

interface TimesheetContextProps {
  listTimesheets: DataTimesheet[];
  totalWorkedHours: string;
  totalOvertimeHours: string;
  totalUndertimeHours: string;
  handleAddTimesheet: (valuesForm: DataTimesheetInput) => void;
  handleUpdateTimesheet: (valuesForm: DataTimesheet) => void;
  handleDeleteTimesheet: (selectedTimesheetId: string) => void;
}

interface TimesheetContextProviderProps {
  children: ReactNode;
}

type FirebaseTimesheets = Record<string, DataTimesheet>;

export const TimesheetContext = createContext<TimesheetContextProps>(
  {} as TimesheetContextProps
);

export function TimesheetContextProvider({
  children,
}: TimesheetContextProviderProps) {
  const { user } = useAuth();
  const [listTimesheets, setListTimesheets] = useState<DataTimesheet[]>([]);
  const [totalWorkedHours, setTotalWorkedHours] = useState("00:00");
  const [totalOvertimeHours, setTotalOvertimeHours] = useState("00:00");
  const [totalUndertimeHours, setTotalUndertimeHours] = useState("00:00");

  useEffect(() => {
    setTotalWorkedHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.workedHours);
      }, "00:00")
    );

    setTotalOvertimeHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.overtimeHours);
      }, "00:00")
    );

    setTotalUndertimeHours(
      listTimesheets.reduce((sumTotal, log) => {
        return sumHours(sumTotal, log.undertimeHours);
      }, "00:00")
    );
  }, [listTimesheets]);

  useEffect(() => {
    const timesheetRef = database.ref();
    timesheetRef.on("value", (timesheet) => {
      const databaseTimesheet = timesheet.val();
      const firebaseTimesheets: FirebaseTimesheets = databaseTimesheet
        ? databaseTimesheet.timesheets
        : {};

      if (firebaseTimesheets) {
        const parsedTimesheets = Object.entries(firebaseTimesheets)
          .map(([key, value]) => {
            return {
              id: key,
              date: value.date,
              idUser: value.idUser,
              arrivalTime: value.arrivalTime,
              lunchExitTime: value.lunchExitTime,
              lunchArrivalTime: value.lunchArrivalTime,
              exitTime: value.exitTime,
              undertimeHours: value.undertimeHours,
              overtimeHours: value.overtimeHours,
              workedHours: value.workedHours,
              createdAt: value.createdAt,
            };
          })
          .filter((timesheet) => timesheet.idUser === user.id);

        setListTimesheets(parsedTimesheets);
      }
    });

    return () => {
      timesheetRef.off("value");
    };
  }, [user.id]);

  const handleAddTimesheet = useCallback(
    async (valuesForm: DataTimesheetInput) => {
      if (isTimesheetExists(listTimesheets, valuesForm)) {
        return errorMessage("You already have a register for this day!");
      }
      const timesheetRef = database.ref("timesheets");
      try {
        await timesheetRef.push({
          ...valuesForm,
          overtimeHours: calcOvertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          undertimeHours: calcUndertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          workedHours: calcWorkedHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          createdAt: String(new Date()),
          date: String(new Date(valuesForm.date)),
          idUser: user.id,
        });
        successMessage("Register created gracefully!");
      } catch (error) {
        errorMessage("Unable to create register!");
      }
    },
    [user.id, listTimesheets]
  );

  const handleUpdateTimesheet = useCallback(
    async (valuesForm: DataTimesheet) => {
      try {
        await database.ref(`timesheets/${valuesForm.id}`).update({
          arrivalTime: valuesForm.arrivalTime,
          lunchExitTime: valuesForm.lunchExitTime,
          lunchArrivalTime: valuesForm.lunchArrivalTime,
          exitTime: valuesForm.exitTime,
          overtimeHours: calcOvertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          undertimeHours: calcUndertimeHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
          workedHours: calcWorkedHours(
            valuesForm.arrivalTime,
            valuesForm.lunchExitTime,
            valuesForm.lunchArrivalTime,
            valuesForm.exitTime
          ),
        });
        successMessage("Register updated gracefully!");
      } catch (error) {
        errorMessage("Unable to update register!");
      }
    },
    []
  );

  const handleDeleteTimesheet = useCallback(
    async (selectedTimesheetId: string) => {
      try {
        await database.ref(`timesheets/${selectedTimesheetId}`).remove();
        successMessage("Register removed gracefully!");
      } catch (error) {
        errorMessage("Unable to remove register!");
      }
    },
    []
  );

  return (
    <TimesheetContext.Provider
      value={{
        listTimesheets,
        totalWorkedHours,
        totalOvertimeHours,
        totalUndertimeHours,
        handleAddTimesheet,
        handleUpdateTimesheet,
        handleDeleteTimesheet,
      }}
    >
      {children}
    </TimesheetContext.Provider>
  );
}
