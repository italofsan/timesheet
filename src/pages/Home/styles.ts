import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { colors } from "../../styles/colors";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: "flex",
      height: "100vh",
      background: "linear-gradient(0.25turn, rgb(143,125,238) , #FFF )",
    },
    appBar: {
      backgroundColor: colors.main,
    },
    title: {
      flexGrow: 1,
    },
    content: {
      flexGrow: 1,
      padding: theme.spacing(3),
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
    },
    imageContainer: {
      display: "flex",
      width: "100%",
      height: "100%",
      padding: "24px 0",
    },
    image: {
      width: "100%",
      maxHeight: 620,
    },
    loginWrapper: {
      display: "flex",
      justifyContent: "center",
    },
    loginContainer: {
      display: "flex",
      flexDirection: "column",
      justifyContent: "center",
      padding: 48,
      maxWidth: 400,
    },
    loginTitle: {
      textAlign: "center",
      fontWeight: "bold",
      fontSize: 32,
      color: colors.main,
      paddingBottom: theme.spacing(3),
    },
    loginSubtitle: {
      paddingBottom: theme.spacing(3),
    },
    loginButton: {
      fontSize: 14,
      fontWeight: "bold",
      padding: "4px 24px",
      color: colors.white,
      transition: ".5s ease",
      backgroundColor: colors.main,
      "&:hover": {
        backgroundColor: colors.mainHover,
      },
      borderRadius: 5,
      textDecoration: "none",
    },
  })
);
