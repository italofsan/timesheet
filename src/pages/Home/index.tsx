import {
  AppBar,
  Container,
  Grid,
  Toolbar,
  Typography,
  Paper,
  Button,
} from "@material-ui/core";
import { useHistory } from "react-router-dom";

import timeHomePageImg from "../../assets/images/time-homepage.svg";

import { useAuth } from "../../hooks/useAuth";

import { useStyles } from "./styles";

export function Home() {
  const classes = useStyles();
  const history = useHistory();
  const { user, signInWithGoogle } = useAuth();

  async function handleLogin() {
    if (!user?.id) {
      await signInWithGoogle();
    }
    history.push("/app");
  }

  return (
    <div className={classes.root}>
      <AppBar position='fixed' className={classes.appBar}>
        <Container maxWidth='lg'>
          <Toolbar>
            <Typography variant='h6' className={classes.title}>
              Timesheet App
            </Typography>
          </Toolbar>
        </Container>
      </AppBar>
      <main className={classes.content}>
        <Container maxWidth='lg'>
          <Grid container justify='center' alignItems='center' direction='row'>
            <Grid item xs={12} sm={6} md={6} lg={6}>
              <div className={classes.imageContainer}>
                <img
                  src={timeHomePageImg}
                  alt='Time Home Page'
                  className={classes.image}
                />
              </div>
            </Grid>
            <Grid
              item
              xs={12}
              sm={6}
              md={6}
              lg={6}
              className={classes.loginWrapper}
            >
              <Paper className={classes.loginContainer}>
                <Typography className={classes.loginTitle}>Log In</Typography>
                <Typography className={classes.loginSubtitle}>
                  To log in to the application you need to use a Google account
                </Typography>
                <Button
                  size='small'
                  variant='contained'
                  className={classes.loginButton}
                  onClick={handleLogin}
                  role='loginButton'
                >
                  Login With Google
                </Button>
              </Paper>
            </Grid>
          </Grid>
        </Container>
      </main>
    </div>
  );
}
