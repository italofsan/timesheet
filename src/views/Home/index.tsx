import { useState } from "react";
import { Button, Container, Grid } from "@material-ui/core";

import { TimesheetTable } from "../../components/TimesheetTable";
import { AddModal } from "../../components/Modals/AddModal";
import { Summary } from "../../components/Summary";

import { useStyles } from "./styles";

export function Home() {
  const classes = useStyles();

  const [addModalOpen, setAddModalOpen] = useState(false);
  function handleOpenAddModal() {
    setAddModalOpen(true);
  }
  function handleCloseAddModal() {
    setAddModalOpen(false);
  }

  return (
    <Container maxWidth='lg' style={{ marginTop: 24 }}>
      <Grid container className={classes.root} spacing={2}>
        <Grid container spacing={4}>
          <Summary />
          <Grid item xs={12} className={classes.buttonContainer}>
            <Button
              size='small'
              variant='contained'
              className={classes.addButton}
              onClick={handleOpenAddModal}
            >
              + Add
            </Button>
          </Grid>
          <TimesheetTable />
        </Grid>

        <AddModal
          addModalOpen={addModalOpen}
          handleCloseAddModal={handleCloseAddModal}
        />
      </Grid>
    </Container>
  );
}
