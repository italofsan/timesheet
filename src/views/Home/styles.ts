import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { colors } from "../../styles/colors";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      height: "100%",
    },
    addButton: {
      fontSize: 14,
      fontWeight: "bold",
      padding: "5px 25px",
      color: colors.white,
      transition: ".5s ease",
      backgroundColor: colors.main,
      "&:hover": {
        backgroundColor: colors.mainHover,
      },
      borderRadius: 5,
      textDecoration: "none",
      [theme.breakpoints.down("xs")]: {
        minWidth: "100%",
      },
    },
    buttonContainer: {
      [theme.breakpoints.down("xs")]: {
        position: "relative",
      },
    },
  })
);
