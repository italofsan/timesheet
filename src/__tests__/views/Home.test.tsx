import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { TimesheetContextProvider } from "../../contexts/TimesheetContext";
import { AuthContextProvider } from "../../contexts/AuthContext";

import { Home } from "../../views/Home";

describe("Home Component", () => {
  it("renders correctly", () => {
    render(
      <AuthContextProvider>
        <TimesheetContextProvider>
          <Home />
        </TimesheetContextProvider>
      </AuthContextProvider>
    );

    expect(screen.getByText("+ Add")).toBeInTheDocument();

    const tree = renderer
      .create(
        <AuthContextProvider>
          <TimesheetContextProvider>
            <Home />
          </TimesheetContextProvider>
        </AuthContextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
