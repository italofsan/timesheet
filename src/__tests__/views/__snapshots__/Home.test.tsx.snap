// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component renders correctly 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
  style={
    Object {
      "marginTop": 24,
    }
  }
>
  <div
    className="MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-4"
      >
        <div
          className="MuiPaper-root makeStyles-workedHoursSummaryContainer-4 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <p
            className="MuiTypography-root makeStyles-workedHoursSummaryValue-6 MuiTypography-body1"
          >
            00:00
          </p>
          <p
            className="MuiTypography-root makeStyles-workedHoursSummaryTitle-5 MuiTypography-body1"
          >
            Total Worked Hours
          </p>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root makeStyles-arrowIcon-12"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
            />
          </svg>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-lg-4"
      >
        <div
          className="MuiPaper-root makeStyles-summaryContainer-7 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <p
            className="MuiTypography-root makeStyles-overtimeHoursSummaryValue-10 MuiTypography-body1"
          >
            00:00
          </p>
          <p
            className="MuiTypography-root makeStyles-summaryTitle-8 MuiTypography-body1"
          >
            Total Overtime Hours
          </p>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root makeStyles-arrowIcon-12"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
            />
          </svg>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-lg-4"
      >
        <div
          className="MuiPaper-root makeStyles-summaryContainer-7 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <p
            className="MuiTypography-root makeStyles-undertimeHoursSummaryValue-11 MuiTypography-body1"
          >
            00:00
          </p>
          <p
            className="MuiTypography-root makeStyles-summaryTitle-8 MuiTypography-body1"
          >
            Total Undertime Hours
          </p>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root makeStyles-arrowIcon-12"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-buttonContainer-3 MuiGrid-item MuiGrid-grid-xs-12"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-addButton-2 MuiButton-containedSizeSmall MuiButton-sizeSmall"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            + Add
          </span>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div
          className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <table
            className="MuiTable-root"
            role={null}
          >
            <caption
              style={
                Object {
                  "backgroundColor": "rgb(143,125,238)",
                  "color": "#FFF",
                }
              }
            >
              You have 0  registers!
            </caption>
            <thead
              className="MuiTableHead-root"
              role={null}
            >
              <tr
                className="MuiTableRow-root MuiTableRow-head"
                role={null}
              >
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Options
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Date
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Arrival
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Lunch Exit
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Lunch Arrival
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Exit
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Worked Hours
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Overtime
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head makeStyles-tableCellHeadOdd-14 MuiTableCell-alignCenter"
                  scope="col"
                >
                  Undertime
                </th>
              </tr>
            </thead>
            <tbody
              className="MuiTableBody-root"
              role={null}
            />
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
